ingenialink/ethercat/network.py

```diff
@@ -31,19 +73,129 @@ class EthercatNetwork:
-        self.servos = []
-        """list: List of the connected servos in the network."""
-        self._ecat_master = None
+    def __init__(
+        self, interface_name: str, connection_timeout: float = DEFAULT_ECAT_CONNECTION_TIMEOUT
+    ):
+        super(EthercatNetwork, self).__init__()
+        self.interface_name: str = interface_name
+        self.servos: List[EthercatServo] = []
+        self.__servos_state: Dict[int, NET_STATE] = {}
+        self.__listener_net_status: Optional[NetStatusListener] = None
+        self.__observers_net_state: Dict[int, List[Any]] = defaultdict(list)
+        self._ecat_master: pysoem.CdefMaster = pysoem.Master()
+        self._ecat_master.open(self.interface_name)
+        self._ecat_master.sdo_read_timeout = int(1_000_000 * connection_timeout)
+        self._ecat_master.sdo_write_timeout = int(1_000_000 * connection_timeout)
+
+    def scan_slaves(self) -> List[int]:
+        """Scans for nodes in the network.
+
+        Returns:
+            List containing all the detected node IDs.
+
+        """
+        nodes = self._ecat_master.config_init()
+        return list(range(1, nodes + 1))
+
+    def connect_to_slave(
+        self,
+        slave_id: int,
+        dictionary: Optional[str] = None,
+        servo_status_listener: bool = False,
+        net_status_listener: bool = False,
+    ) -> EthercatServo:
+        """Connects to a drive through a given slave number.
+
+        Args:
+            slave_id: Targeted slave to be connected.
+            dictionary: Path to the dictionary file.
+            servo_status_listener: Toggle the listener of the servo for
+                its status, errors, faults, etc.
+            net_status_listener: Toggle the listener of the network
+                status, connection and disconnection.
+
+        Raises:
+            ValueError: If the slave ID is not valid.
+            ILError: If no slaves are found.
+
+        """
+        if not isinstance(slave_id, int) or slave_id < 0:
+            raise ValueError("Invalid slave ID value")
+        slaves = self.scan_slaves()
+        if len(slaves) == 0:
+            raise ILError("Could not find any slaves in the network.")
+        if slave_id not in slaves:
+            raise (ILError(f"Slave {slave_id} was not found."))
+        slave = self._ecat_master.slaves[slave_id - 1]
+        servo = EthercatServo(slave, slave_id, dictionary, servo_status_listener)
+        self.servos.append(servo)
+        self._set_servo_state(slave_id, NET_STATE.CONNECTED)
+        if net_status_listener:
+            self.start_status_listener()
+        return servo
+
+    def disconnect_from_slave(self, servo: EthercatServo) -> None:
+        """Disconnects the slave from the network.
+
+        Args:
+            servo: Instance of the servo connected.
+
+        """
+        self.servos.remove(servo)
+        if not self.servos:
+            self.stop_status_listener()
+            self._ecat_master.close()
```

Comment:
If you connect a drive, then disconnect it and try to reconnect a very ugly error appears.
Process finished with exit code -1073741819 (0xC0000005)

Reply:
Done